tools/add_pre_post_processing_to_model.py

```diff
@@ -0,0 +1,218 @@
+    model_output_shape = model.graph.output[0].type.tensor_type.shape
+    assert model_input_shape.dim[-1].HasField("dim_value")
+    assert model_input_shape.dim[-2].HasField("dim_value")
+    assert model_output_shape.dim[-1].HasField("dim_value")
+    assert model_output_shape.dim[-2].HasField("dim_value")
+
+    w_in = model_input_shape.dim[-1].dim_value
+    h_in = model_input_shape.dim[-2].dim_value
+    h_out = model_output_shape.dim[-2].dim_value
+    w_out = model_output_shape.dim[-1].dim_value
+
+    # pre/post processing for https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html
+    pipeline = PrePostProcessor(inputs)
+    pipeline.add_pre_processing(
+        [
+            ConvertImageToBGR(),  # jpg/png image to BGR in HWC layout
+            Resize((h_in, w_in)),
+            CenterCrop(h_in, w_in),
+            # this produces Y, Cb and Cr outputs. each has shape {h_in, w_in}. only Y is input to model
+            PixelsToYCbCr(layout="BGR"),
+            # if you inserted this Debug step here the 3 outputs from PixelsToYCbCr would also be model outputs
+            # Debug(num_inputs=3),
+            ImageBytesToFloat(),  # Convert Y to float in range 0..1
+            Unsqueeze([0, 1]),  # add batch and channels dim to Y so shape is {1, 1, h_in, w_in}
+        ]
+    )
+
+    # Post-processing is complicated here. resize the Cb and Cr outputs from the pre-processing to match
+    # the model output size, merge those with the Y` model output, and convert back to RGB.
+
+    # create the Steps we need to use in the manual connections
+    pipeline.add_post_processing(
+        [
+            Squeeze([0, 1]),  # remove batch and channels dims from Y'
+            FloatToImageBytes(name="Yout_to_bytes"),  # convert Y' to uint8 in range 0..255
+            # Resize the Cb values (output 1 from PixelsToYCbCr)
+            (
+                Resize((h_out, w_out), "HW"),
+                [IoMapEntry(producer="PixelsToYCbCr", producer_idx=1, consumer_idx=0)],
+            ),
+            # the Cb and Cr values are already in the range 0..255 so multiplier is 1. we're using the step to round
+            # for accuracy (a direct Cast would just truncate) and clip (to ensure range 0..255) the values post-Resize
+            FloatToImageBytes(multiplier=1.0, name="Resized_Cb"),
+            (Resize((h_out, w_out), "HW"), [IoMapEntry("PixelsToYCbCr", 2, 0)]),
+            FloatToImageBytes(multiplier=1.0, name="Resized_Cr"),
+            # as we're selecting outputs from multiple previous steps we need to map them to the inputs using step names
+            (
+                YCbCrToPixels(layout="BGR"),
+                [
+                    IoMapEntry("Yout_to_bytes", 0, 0),  # uint8 Y' with shape {h, w}
+                    IoMapEntry("Resized_Cb", 0, 1),  # uint8 Cb'
+                    IoMapEntry("Resized_Cr", 0, 2),
+                ],
+            ),  # uint8 Cr'
```

Comment:
looks like this comment is on the wrong line

Reply:
The joys of 'black'. Most of its formatting is good. Some is dubious.